### PR TITLE
feat: Add devcontainer config

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

* Add devcontainer config using the pyhf Dockerfile.
   - c.f. https://containers.dev/
   - Motivated by @crazy4pi314's SciPy 2023 lightning talk.

@crazy4pi314, @guenp, and @ltalirz also gave the SciPy 2023 tutorial [Meet your coding best friend: VS Code💖 - A hands-on tutorial on how to get the most out of the world’s most popular Python editor](https://cfp.scipy.org/2023/talk/RKV3PZ/) which covers dev containers in it as well: https://github.com/crazy4pi314/scipy-vscode-tutorial

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add devcontainer config using the pyhf Dockerfile.
   - c.f. https://containers.dev/
   - Motivated by Sarah Kaiser's SciPy 2023 lightning talk.
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
:wave: @crazy4pi314 following your great :zap: talk today and that things seemed pretty simple in https://github.com/bluesky/tiled/pull/522 I wanted to try to get this going for `pyhf` as well. I'll read https://containers.dev/ but if you also have time to discuss dev containers (or review this PR) before you leave SciPy 2023 that would be great! :)

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	"customizations": {
+		"vscode": {
+			"settings": {
+				"python.defaultInterpreterPath": "/opt/venv/bin/python"
+			},
+			"extensions": [
+				"ms-python.python",
+				"ms-toolsai.jupyter",
+				"GitHub.codespaces",
+				"ms-vscode-remote.remote-containers"
```
Comment (ltalirz) [inline .devcontainer/devcontainer.json:20]: I'll let @crazy4pi314 give the final answer here, but my understanding is that those extensions are not actually necessary here, unless you want to connect to a remote container / codespace from _inside_ the dev container (of course the extensions should not hurt either).

Comment (matthewfeickert) [reply .devcontainer/devcontainer.json:20]: > unless you want to connect to a remote container / codespace

While the pyhf dev team doesn't actively use Codespaces at the moment (very cool, but not clear if the cost is worth it atm for us) we would like for others to use Codespaces or GitPod for demos and to help people making contributions get onboarded. Do these extensions help / become needed there?

Comment (crazy4pi314) [reply .devcontainer/devcontainer.json:20]: Yeah I think they are probably not needed, Its likely my fault that I was exporting envs and those got installed as needed (but you don't have to proactively)

**Comment by matthewfeickert** [comment]
Thank you @crazy4pi314 and @ltalirz! :rocket:

**Pull request merged**